### PR TITLE
Adds DcLoadFlowContext

### PR DESCRIPTION
**Pull request by jlbouchot**
Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

This PR adds the DcLoadFlowContext class and propagates it into the currently existing code base. This context simplifies handling DcEngine runs

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by jlbouchot** [comment]
@annetill , @geofjamg , @obrix : a couple of things to look at: 1) I have kept the DcTargetVector class but pretty much commented it everywhere. Do you think we should keep it for future developments? 2) Some code can still be refactor within the DcLoadFlowEngine class (e.g. avoiding passing jacobian as parameters to functions down and down again). Should this be done now?

File: src/main/java/com/powsybl/openloadflow/dc/DcTargetVector.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import com.powsybl.openloadflow.equations.Equation;
+import com.powsybl.openloadflow.equations.EquationSystem;
+import com.powsybl.openloadflow.equations.EquationTerm;
+import com.powsybl.openloadflow.equations.TargetVector;
+import com.powsybl.openloadflow.network.*;
+
+/**
+ * @author Jean-Luc Bouchot (Artelys) <jlbouchot at gmail.com>
+ */
+public class DcTargetVector extends TargetVector<DcVariableType, DcEquationType> {
```
Comment (geofjamg) [inline src/main/java/com/powsybl/openloadflow/dc/DcTargetVector.java:20]: So you can remove the method initTarget of DcLoadFlowEngine?

File: src/main/java/com/powsybl/openloadflow/dc/DcValueVoltageInitializer.java [hunk 2]
```diff
@@ -57,8 +57,8 @@ public void prepare(LfNetwork network) {
                                                                    distributedSlack,
                                                                    balanceType,
                                                                    false);
-        DcLoadFlowEngine engine = new DcLoadFlowEngine(List.of(network), parameters);
-        if (engine.run(reporter).get(0).getStatus() != LoadFlowResult.ComponentResult.Status.CONVERGED) {
+        DcLoadFlowEngine engine = new DcLoadFlowEngine(new DcLoadFlowContext(network, parameters));
```
Comment (geofjamg) [inline src/main/java/com/powsybl/openloadflow/dc/DcValueVoltageInitializer.java:60]: DcLoadFlowContext has to be closed

File: src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowContext.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import com.powsybl.openloadflow.equations.JacobianMatrix;
+// import com.powsybl.openloadflow.equations.TargetVector;
+import com.powsybl.openloadflow.network.LfNetwork;
+
+import java.util.Objects;
+
+/**
+ * @author Jean-Luc Bouchot (Artelys) <jlbouchot at gmail.com>
+ */
+public class DcLoadFlowContext implements AutoCloseable {
```
Comment (geofjamg) [inline src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowContext.java:23]: Maybe we should add also an `AbstractLoadFlowContext` to share `LfNetwork` and `JacobianMatrix`

**Comment by geofjamg** [comment]
> @annetill , @geofjamg , @obrix : a couple of things to look at: 1) I have kept the DcTargetVector class but pretty much commented it everywhere. Do you think we should keep it for future developments? 2) Some code can still be refactor within the DcLoadFlowEngine class (e.g. avoiding passing jacobian as parameters to functions down and down again). Should this be done now?

1) Except maybe for usage of DcLoadFlowEngine in sensi, we should rely only on new DcTargetVector (LF and SA).
2) be careful to not change to much thing for the sensi analysis, the run static methods has been done for this purpose.

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowContext.java (hunk 3 above)
Comment (jlbouchot) [reply src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowContext.java:23]: I agree with this and we actually thought about it. This raises a question though in terms of files within the project. Today, DcLoadFlowContext is found within the OLF.dc package while the AcLoadFlowContext is found within the OLF.ac.outerloop package. Should we have an extra "abstractflow" package which would be implemented within their respective AC/DC packages?

**Review by geofjamg — APPROVED** [review]

File: src/test/java/com/powsybl/openloadflow/dc/DcLoadFlowTest.java [hunk 4]
```diff
@@ -259,4 +259,11 @@ void shuntCompensatorActivePowerZero() {
         loadFlowRunner.run(network, parameters);
         LoadFlowAssert.assertActivePowerEquals(0, sc.getTerminal());
     }
+
+    @Test
+    void dcLoadFlowContextTest() {
```
Comment (annetill) [inline src/test/java/com/powsybl/openloadflow/dc/DcLoadFlowTest.java:264]: This test is definitive?

File: src/main/java/com/powsybl/openloadflow/lf/LoadFlowResult.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Copyright (c) 2022, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/lf/LoadFlowResult.java:5]: Can I add the 'SPDX-License-Identifier: MPL-2.0` everywhere?

File: src/test/java/com/powsybl/openloadflow/dc/DcLoadFlowTest.java (hunk 4 above)
Comment (geofjamg) [reply src/test/java/com/powsybl/openloadflow/dc/DcLoadFlowTest.java]: Removed

File: src/main/java/com/powsybl/openloadflow/lf/LoadFlowResult.java (hunk 5 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/lf/LoadFlowResult.java:5]: Added

**Pull request merged**